### PR TITLE
Fix stackoverflow issue in ballerina shell when execute main()

### LINKED ISSUE #41419: [Bug]: StackOverflow error occurs when executing `main()` in Ballerina Shell
### Description

When attempting to execute the `main()` function in the Ballerina shell, it attempts to call the main function recursively as shown below. A proper error message should be displayed when the user enters `main()`.

```
panic: {ballerina}StackOverflow {}
at main()
at main()
at main()
at main()
at main()
at main()
...
```

### Steps to Reproduce

Open the Ballerina shell, and execute `main()` as an input.

### Affected Version(s)

Ballerina 2201.8.0 (Swan Lake Update 8)

### OS, DB, other environment details and versions

OS: macOS 13.5.2
Java: openjdk version "17.0.8" 2023-07-18

### Related area

-> Shell

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by lwpamihiranga**
Fixes #41419 

- Add mechanism to check if source includes restricted keywords

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: ballerina-shell/modules/shell-core/src/main/java/io/ballerina/shell/parser/ParserConstants.java [hunk 1]
```diff
@@ -42,5 +45,16 @@ public static boolean isFunctionNameRestricted(String functionName) {
...
 
+    /**
+     * Check if the given source includes restricted keywords
+     * @param source
+     * @return {@code Optional<String>} if the source include restricted keywords. {@code Optional<Empty>} otherwise
+     */
+    public static Optional<String> getMatchedRestrictedSource(String source) {
+        return RESTRICTED_SOURCE_NAMES.stream()
+                .filter(source::contains)
+                .findFirst();
```
Comment (LakshanWeerasinghe) [inline ballerina-shell/modules/shell-core/src/main/java/io/ballerina/shell/parser/ParserConstants.java:56]:
```suggestion
        return RESTRICTED_SOURCE_NAMES.contains(source);
```

File: ballerina-shell/modules/shell-core/src/main/java/io/ballerina/shell/parser/ParserConstants.java [hunk 2]
```diff
@@ -42,5 +45,16 @@ public static boolean isFunctionNameRestricted(String functionName) {
         return RESTRICTED_FUNCTION_NAMES.contains(functionName) || functionName.startsWith(WRAPPER_PREFIX);
     }
 
+    /**
+     * Check if the given source includes restricted keywords
+     * @param source
+     * @return {@code Optional<String>} if the source include restricted keywords. {@code Optional<Empty>} otherwise
+     */
+    public static Optional<String> getMatchedRestrictedSource(String source) {
```
Comment (LakshanWeerasinghe) [inline ballerina-shell/modules/shell-core/src/main/java/io/ballerina/shell/parser/ParserConstants.java:53]:
```suggestion
    public static boolean isRestrictedSource(String source) {
```

**Review by LakshanWeerasinghe — COMMENTED** [review]
There are few check style failures fix them. And add tests.

File: ballerina-shell/modules/shell-core/src/main/java/io/ballerina/shell/parser/SerialTreeParser.java [hunk 3]
```diff
@@ -130,4 +134,19 @@ private boolean isModuleDeclarationAllowed(ModuleMemberDeclarationNode declarati
...
+     */
+    private boolean isSourceAllowed(String source) {
+        Optional<String> matchedSource = ParserConstants.getMatchedRestrictedSource(source);
+
+        if (matchedSource.isPresent()) {
+            addWarnDiagnostic("Found restricted keyword '" + matchedSource.get() + "' in the declaration.\n" +
+                    "Discarded '" + source + "' without loading.");
+            return false;
+        }
+
```
Comment (SasinduDilshara) [inline ballerina-shell/modules/shell-core/src/main/java/io/ballerina/shell/parser/SerialTreeParser.java:149]: I would prefer to add new lines after a if/for condition with huge length

File: ballerina-shell/modules/shell-core/src/main/java/io/ballerina/shell/parser/ParserConstants.java (hunk 1 above)
Comment (lwpamihiranga) [reply ballerina-shell/modules/shell-core/src/main/java/io/ballerina/shell/parser/ParserConstants.java:56]: Hi @LakshanWeerasinghe,

Thanks for the review. This was my first approach as well, returning a boolean value. However, if we were to return a boolean value we wouldn't be able to provide a detailed error message explaining why the input is rejected, as in  [this code block](https://github.com/ballerina-platform/ballerina-lang/pull/43058/files#diff-c7cfdebbd1741dc7606c90eeeb71bf67db2686e5243609cce468d04fd0c3dc9fR145-R146). What do you think?

See below for more reference.

> Found restricted keyword '_man()_' in the declaration.

bal-shell-stackoverflow-main-fix

File: ballerina-shell/modules/shell-core/src/main/java/io/ballerina/shell/parser/SerialTreeParser.java (hunk 3 above)
Comment (lwpamihiranga) [reply ballerina-shell/modules/shell-core/src/main/java/io/ballerina/shell/parser/SerialTreeParser.java]: Thanks for the suggestion Sasindu. I removed the newline.

**Comment by SasinduDilshara** [comment]
Few LS related tests are failing in this PR

**Comment by SasinduDilshara** [comment]
PR contains a build failure in the distribution

**Review by nipunayf — COMMENTED** [review]
Thanks for the PR; it's a good start for the fix. However, there are a few minor concerns with your changes.  

First, the bug still exists when there is whitespace between the tokens (e.g., `main   ()`). Also, introducing a new `RESTRICTED_SOURCE_NAMES` isn't necessary. We can reuse the `RESTRICTED_FUNCTION_NAMES` since the proposed solution should apply to other related functions as well. Currently, we only use `RESTRICTED_FUNCTION_NAMES` for function definitions, so for this fix, we need to reuse it for function calls.

**Comment by lwpamihiranga** [comment]
> Thanks for the PR; it's a good start for the fix. However, there are a few minor concerns with your changes.
> 
> First, the bug still exists when there is whitespace between the tokens (e.g., `main ()`). Also, introducing a new `RESTRICTED_SOURCE_NAMES` isn't necessary. We can reuse the `RESTRICTED_FUNCTION_NAMES` since the proposed solution should apply to other related functions as well. Currently, we only use `RESTRICTED_FUNCTION_NAMES` for function definitions, so for this fix, we need to reuse it for function calls.

Thanks for the insightful review Nipuna. I have updated my solution to check the expressions in the parser level using the existing `RESTRICTED_FUNCTION_NAMES`.

File: ballerina-shell/modules/shell-core/src/main/java/io/ballerina/shell/parser/trials/ExpressionTrial.java [hunk 4]
```diff
@@ -57,7 +58,16 @@ public Collection<Node> parse(String source) throws ParserTrialFailedException {
...
         }
+        validateExpression(expressionNode.toSourceCode());
         nodes.add(expressionNode);
         return nodes;
     }
+
+    private void validateExpression(String expression) {
+        String functionName = expression.replaceAll("\\s*\\(.*", "");
+        if (ParserConstants.isFunctionNameRestricted(functionName)) {
+            String message = "Function name " + "'" + functionName + "'" + " not allowed in Ballerina Shell.\n";
```
Comment (nipunayf) [inline ballerina-shell/modules/shell-core/src/main/java/io/ballerina/shell/parser/trials/ExpressionTrial.java:69]:
```suggestion
            String message = String.format("Function name '%s' not allowed in Ballerina Shell.%n", functionName);
```

**Review by nipunayf — COMMENTED** [review]
Shall we have a test case for this in [`errors.parse.json`](https://github.com/ballerina-platform/ballerina-lang/blob/64979464bde488ef4e6ca10046467d3f10ff8f2c/misc/ls-extensions/modules/bal-shell-service/src/test/resources/testcases/getResult/errors.parse.json)?

File: ballerina-shell/modules/shell-core/src/main/java/io/ballerina/shell/parser/trials/ExpressionTrial.java [hunk 5]
```diff
@@ -57,7 +58,16 @@ public Collection<Node> parse(String source) throws ParserTrialFailedException {
...
         }
+        validateExpression(expressionNode.toSourceCode());
         nodes.add(expressionNode);
         return nodes;
     }
+
+    private void validateExpression(String expression) {
+        String functionName = expression.replaceAll("\\s*\\(.*", "");
+        if (ParserConstants.isFunctionNameRestricted(functionName)) {
+            String message = String.format("Function name '%s' not allowed in Ballerina Shell.\n", functionName);
```
Comment (nipunayf) [inline ballerina-shell/modules/shell-core/src/main/java/io/ballerina/shell/parser/trials/ExpressionTrial.java:69]:
```suggestion
            String message = String.format("Function name '%s' not allowed in Ballerina Shell.%n", functionName);
```

**Comment by lwpamihiranga** [comment]
> Shall we have a test case for this in [`errors.parse.json`](https://github.com/ballerina-platform/ballerina-lang/blob/64979464bde488ef4e6ca10046467d3f10ff8f2c/misc/ls-extensions/modules/bal-shell-service/src/test/resources/testcases/getResult/errors.parse.json)?

I have added the test.

**Review by nipunayf — APPROVED** [review]
The changes LGTM. Shall we rebase the PR against the master branch so we can proceed with the merge?

**Comment by lwpamihiranga** [comment]
> The changes LGTM. Shall we rebase the PR against the master branch so we can proceed with the merge?

Thank you! I have merged the newest changes to this feature

**Comment by nipunayf** [comment]
The PR is now merged, thanks for the contribution @lwpamihiranga 🎉

**Comment by lwpamihiranga** [comment]
> The PR is now merged, thanks for the contribution @lwpamihiranga 🎉

Thank you for your help and guidance, @nipunayf. Do you have any suggestions for an issue I could tackle next?

**Comment by nipunayf** [comment]
> > The PR is now merged, thanks for the contribution @lwpamihiranga 🎉
> 
> Thank you for your help and guidance, @nipunayf. Do you have any suggestions for an issue I could tackle next?

Thank you for your interest in the project! To get started, you can browse through the "good first issues" section: https://github.com/ballerina-platform/ballerina-lang/issues?q=is%3Aopen+is%3Aissue+label%3A%22good+first+issue%22. I suggest starting with https://github.com/ballerina-platform/ballerina-lang/issues/35018, which should be an easy issue to tackle.

**Pull request merged**